src/plugins/mpath.c

```diff
@@ -156,360 +184,295 @@ gboolean bd_mpath_is_tech_avail (BDMpathTech tech, guint64 mode, GError **error)
+        for (j = 0; j < pg_count; ++j) {
+            dmmp_path_array_get (pgs[j], &ps, &p_count);
+            for (l = 0; l < p_count; ++l) {
+                blk_name = dmmp_path_blk_name_get (ps[l]);
+                if (g_strcmp0 (blk_name, dev_name) == 0) {
+                    rc = TRUE;
+                    break;
+                }
             }
-            for (dev_name = deps; !ret && *dev_name; dev_name++)
-                ret = (g_strcmp0 (*dev_name, device) == 0);
-            g_strfreev (deps);
-        } else if (*error) {
-            g_prefix_error (error, "Failed to determine map's target for '%s'", names->name);
-            g_free (dev_path);
-            dm_task_destroy (task_names);
-            return FALSE;
         }
-    } while (!ret && next);
+    }
 
-    g_free (dev_path);
-    dm_task_destroy (task_names);
-    return ret;
+out:
+    dmmp_context_free(ctx);
+    if (dmmp_rc != DMMP_OK)
+        g_set_error (error, BD_MPATH_ERROR, BD_MPATH_ERROR_FLUSH,
+                     "%d: %s", dmmp_rc, _dmmp_last_err_msg(ctx));
+    if (dev_path != NULL)
+        g_free(dev_path);
+    if (dev_name != NULL)
+        g_free(dev_name);
+    return rc;
 }
 
 /**
  * bd_mpath_get_mpath_members:
  * @error: (out): place to store error (if any)
  *
- * Returns: (transfer full) (array zero-terminated=1): list of names of all devices that are
- *                                                     members of the mpath mappings
- *                                                     (or %NULL in case of error)
+ * Returns: (transfer full) (array zero-terminated=1): list of names of all
+ * devices that are members of the mpath mappings (or %NULL in case of error)
  *
  * Tech category: %BD_MPATH_TECH_BASE-%BD_MPATH_TECH_MODE_QUERY
  */
 gchar** bd_mpath_get_mpath_members (GError **error) {
-    struct dm_task *task_names = NULL;
-	struct dm_names *names = NULL;
-    guint64 next = 0;
-    gchar **deps = NULL;
-    gchar **dev_name = NULL;
-    guint64 n_deps = 0;
-    guint64 n_devs = 0;
-    guint64 top_dev = 0;
+    struct dmmp_context *ctx = NULL;
+    struct dmmp_mpath **mps = NULL;
+    uint32_t mp_count = 0;
+    uint32_t i = 0;
+    uint32_t j = 0;
+    uint32_t l = 0;
+    struct dmmp_path_group **pgs = NULL;
+    uint32_t pg_count = 0;
+    struct dmmp_path **ps = NULL;
+    uint32_t p_count = 0;
+    int dmmp_rc = DMMP_OK;
+    const char *blk_name = NULL;
     gchar **ret = NULL;
-    guint64 progress_id = 0;
-
-    progress_id = bd_utils_report_started ("Started getting mpath members");
-
-    if (geteuid () != 0) {
-        g_set_error (error, BD_MPATH_ERROR, BD_MPATH_ERROR_NOT_ROOT,
-                     "Not running as root, cannot query DM maps");
-        bd_utils_report_finished (progress_id, (*error)->message);
-        return NULL;
+    GList *path_blk_names = NULL;
+    gchar *dup = NULL;
+    guint len = 0;
+    guint m = 0;
+    GList *cur_node = NULL;
```

Comment:
Please use a ``GPtrArray`` here. And note that ``g_ptr_array_free (arr, FALSE)`` can be used to free the container while preserving the content which means the content doesn't have to be copied over elsewhere.

Reply:
The memory holding the path device string is managed/owned by `struct dmmp_mpath` where we cannot only partly free(free all memory except the path device string). Hence `strdup()` is required.